src/fingerprinting/hooks.ts

```diff
@@ -0,0 +1,88 @@
+import { Fingerprint, FingerprintInjector } from 'fingerprint-injector';
+import { BrowserPool, PlaywrightPlugin, PuppeteerPlugin } from '..';
+import { BrowserController } from '../abstract-classes/browser-controller';
+import { LaunchContext } from '../launch-context';
+import { getGeneratorDefaultOptions } from './utils';
+
+// @ts-expect-error I need help how to fix this :/
+export const createFingerprintPreLaunchHook = (browserPool: BrowserPool) => {
+    const {
+        fingerprintGenerator,
+        fingerprintCache,
+        fingerprintsOptions: {
+            fingerprintGeneratorOptions,
+        },
+    } = browserPool;
+
+    return (_pageId: string, launchContext: LaunchContext) => {
+        const { useIncognitoPages, launchOptions, proxyUrl } = launchContext;
+        // If no options are passed we try to pass best default options as possible to match browser and OS.
+        const fingerprintGeneratorFinalOptions = fingerprintGeneratorOptions || getGeneratorDefaultOptions(launchContext);
+        let fingerprint;
+
+        if (proxyUrl && fingerprintCache?.has(proxyUrl)) {
+            fingerprint = fingerprintCache.get(proxyUrl);
+        } else {
+            fingerprint = fingerprintGenerator.getFingerprint(fingerprintGeneratorFinalOptions).fingerprint;
+            fingerprintCache?.set(proxyUrl, fingerprint);
```

Comment:
once you fix the method signature, you will see that `proxyUrl` can be undefined, not sure what the logic here is, if you are sure it will be defined, use `proxyUrl!`, if not, add it to the `else if` condition

Reply:
It is the per proxy URL fingerprint caching. We want to use the same fingerprint per IP. However, this is only valid if the `proxyUrl` string is unique per IP. I will improve the condition - there is no reason to manipulate the cache if there is no `proxyUrl`.